src/Algebra/Graph/AdjacencyIntMap/Algorithm.hs

```diff
@@ -213,32 +239,86 @@ dfs vs = concatMap flatten . dfsForestFrom vs
+    exit v = modify' (\(S p vs) -> S (IntMap.alter (fmap leave) v p) (v:vs))
+      where leave = \case
+              Entered x -> Exited x
+              _ -> error $ "Internal error: dfs search order violated"
+    retrace curr head table = aux (curr :| []) where
+      aux xs@(curr :| _)
+        | head == curr = xs
+        | otherwise = case table IntMap.! curr of
+            Entered prev -> aux (prev <| xs)
+            _ -> error "Internal error: dfs back edge mistakingly identified"
 
--- | Compute the /topological sort/ of a graph or return @Nothing@ if the graph
--- is cyclic.
+-- | Compute a topological sort of a DAG or discover a cycle.
+--
+--   Vertices are expanded in increasing order with respect to their
+--   'Ord' instance. This gives the lexicographically smallest
+--   topological ordering in the case of success. In the case of
+--   failure, the cycle is characterized by being the
+--   lexicographically smallest up to rotation with respect to @Ord
+--   (Dual Int)@ in the first connected component of the graph
+--   containing a cycle, where the connected components are ordered by
+--   their largest vertex with respect to @Ord a@.
+--
+--   Complexity: /O((n+m)*min(n,W))/ time and /O(n)/ space.
 --
 -- @
--- topSort (1 * 2 + 3 * 1)               == Just [3,1,2]
--- topSort (1 * 2 + 2 * 1)               == Nothing
--- fmap ('flip' 'isTopSortOf' x) (topSort x) /= Just False
--- 'isJust' . topSort                      == 'isAcyclic'
+-- topSort (1 * 2 + 3 * 1)                    == Right [3,1,2]
+-- topSort ('path' [1..5])                      == Right [1..5]
+-- topSort (3 * (1 * 4 + 2 * 5))              == Right [3,1,2,4,5]
+-- topSort (1 * 2 + 2 * 1)                    == Left (2 ':|' [1])
+-- topSort ('path' [5,4..1] + 'edge' 2 4)         == Left (4 ':|' [3,2])
+-- topSort ('circuit' [1..3])                   == Left (3 ':|' [1,2])
+-- topSort ('circuit' [1..3] + 'circuit' [3,2,1]) == Left (3 ':|' [2])
+-- topSort (1*2 + 2*1 + 3*4 + 4*3 + 5*1)      == Left (1 ':|' [2])
+-- fmap ('flip' 'isTopSortOf' x) (topSort x)      /= Right False
+-- 'isRight' . topSort                          == 'isAcyclic'
 -- @
-topSort :: AdjacencyIntMap -> Maybe [Int]
-topSort m = if isTopSortOf result m then Just result else Nothing
-  where
-    result = Typed.topSort (Typed.fromAdjacencyIntMap m)
+topSort :: AdjacencyIntMap -> Either (NonEmpty Int) [Int]
+topSort g = runContT (evalStateT (topSort' g) (S IntMap.empty [])) id 
 
 -- | Check if a given graph is /acyclic/.
 --
+--   Complexity: /O((n+m)*min(n,W))/ time and /O(n)/ space.
+--
 -- @
 -- isAcyclic (1 * 2 + 3 * 1) == True
 -- isAcyclic (1 * 2 + 2 * 1) == False
 -- isAcyclic . 'circuit'       == 'null'
--- isAcyclic                 == 'isJust' . 'topSort'
+-- isAcyclic                 == 'isRight' . 'topSort'
 -- @
 isAcyclic :: AdjacencyIntMap -> Bool
-isAcyclic = isJust . topSort
+isAcyclic g = case topSort g of
```

Comment:
I think this should be just `isRight . topSort`:

https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Either.html#v:isRight

Reply:
I went with `isRight` at first, but I think it caused some of the earlier GHCs to fail the CI tests

Reply:
never mind! not sure what went wrong before!